Objects/unicodeobject.c

```diff
@@ -2217,6 +2217,18 @@ PyUnicode_FromWideChar(const wchar_t *u, Py_ssize_t size)
     if (size == 0)
         _Py_RETURN_UNICODE_EMPTY();
 
+#ifdef HAVE_NON_UNICODE_WCHAR_T_REPRESENTATION
+    /* Oracle Solaris uses non-Unicode internal wchar_t form for
+       non-Unicode locales and hence needs conversion to UTF first. */
+    char* codeset = nl_langinfo(CODESET);
+    if (strcmp(codeset, "UTF-8") && strcmp(codeset, "646")) {
```

Comment:
```suggestion
    if (strcmp(codeset, "UTF-8") != 0 && strcmp(codeset, "646") != 0) {
```

You can maybe add an helper function in fileutils.c to implement nl_langinfo() + strcmp() check. Something like _Py_IsLocaleUnicodeWchar(). So you can put it near _Py_convert_wchar_t_to_UTF32() implementation.